spock-core/src/main/java/org/spockframework/compiler/ConditionRewriter.java

```diff
@@ -486,13 +495,33 @@ private Expression unrecord(Expression expr) {
     return ((ArgumentListExpression)methodExpr.getArguments()).getExpression(1);
   }
 
+  // extractVariableNumber(record(expr)) ==
+  // extractVariableNumber($spock_valueRecorder.record($spock_valueRecorder.startRecordingValue(variableNumber), expr) ==
+  // variableNumber
+  private int extractVariableNumber(Expression expr){
+    if (!(expr instanceof MethodCallExpression)) return -1;
+    MethodCallExpression methodExpr = (MethodCallExpression) expr;
+    Expression targetExpr = methodExpr.getObjectExpression();
+    if (!(targetExpr instanceof VariableExpression)) return -1;
+    VariableExpression var = (VariableExpression)targetExpr;
+    if (!var.getName().equals("$spock_valueRecorder")) return -1;
```

Comment:
equality on strings should always be "constant".equals(variable)
`!"$spock_valueRecorder".equals(var.getName())`